### PR TITLE
[Impeller] Start a more generic stroke tessellator for path

### PR DESCRIPTION
**Pull request by dnfield**
This is the start of a more generic tessellator for strokes.

Still missing are:

- Support for multiple contours in a single path
- Support for non-default endcaps.
- Support for non-default line joins.

Some other things that are still left to do here:
- Better tuning around buffer allocation. Right now I'm just allocating buffers that should be oversized for typical paths seen in Flutter. This won't be too good long term.

It does now support arbitrarily mixed path data and ingestion from a `Path` object. I've tested it on one relatively simple path.

### CODE REVIEW

**Comment by dnfield** [comment]
Argh, I had not rerun gn check on this, I'll fix up the build rules.

File: impeller/renderer/compute_tessellator.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+#include "impeller/geometry/path.h"
+#include "impeller/renderer/buffer_view.h"
+#include "impeller/renderer/command_buffer.h"
+#include "impeller/renderer/context.h"
+
+namespace impeller {
+
+//------------------------------------------------------------------------------
+/// @brief      A utility that generates triangles of the specified fill type
+///             given a polyline. This happens on the CPU.
```
Comment (jonahwilliams) [inline impeller/renderer/compute_tessellator.h:17]: nit: what happens on the CPU? Polyline generation right?

File: impeller/renderer/compute_tessellator.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  ComputeTessellator& SetStrokeJoin(Join value);
+  ComputeTessellator& SetStrokeCap(Cap value);
+  ComputeTessellator& SetMiterLimit(Scalar value);
+  ComputeTessellator& SetCubicAccuracy(Scalar value);
+  ComputeTessellator& SetQuadraticTolerance(Scalar value);
+
+  //----------------------------------------------------------------------------
+  /// @brief      Generates triangles from the path.
+  ///
+  /// @return  A buffer view that will be filled with the tessellation result.
```
Comment (jonahwilliams) [inline impeller/renderer/compute_tessellator.h:53]: Seems to return a status?

File: impeller/renderer/compute_tessellator.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  ///
+  Status Tessellate(
```
Comment (jonahwilliams) [inline impeller/renderer/compute_tessellator.h:55]: How does the synchronization work here? I give you a buffer view then once the completion callback has fired then this particular compute workload is done right?

**Review by jonahwilliams — APPROVED** [review]
Lgtm with some questions?

File: impeller/renderer/compute_tessellator.h (hunk 1 above)
Comment (dnfield) [reply impeller/renderer/compute_tessellator.h:17]: I'll fix this, it's an unintentional miss from copying and pasting.

File: impeller/compiler/compiler.cc
Comment (dnfield) [inline impeller/compiler/compiler.cc:1]: @chinmaygarde please review this

Comment (dnfield) [reply impeller/compiler/compiler.cc:1]: (Particularly for the Vulkan change)

File: impeller/tools/malioc.json [hunk 4]
```diff
@@ -13266,6 +13263,131 @@
       }
     }
   },
+  "flutter/impeller/renderer/path_polyline.comp.vkspv": {
```
Comment (dnfield) [inline impeller/tools/malioc.json:13266]: I have no idea why the changes above are happening. I tried commenting out chagnes in compiler.cc and they still were there.

File: impeller/playground/backend/metal/playground_impl_mtl.mm [hunk 5]
```diff
@@ -34,23 +35,25 @@
...
+              impeller_fixtures_shaders_data, impeller_fixtures_shaders_length),
+          std::make_shared<fml::NonOwnedMapping>(
+              impeller_subgroup_fixtures_shaders_data,
+              impeller_subgroup_fixtures_shaders_length),
+          std::make_shared<fml::NonOwnedMapping>(impeller_imgui_shaders_data,
+                                                 impeller_imgui_shaders_length),
+          std::make_shared<fml::NonOwnedMapping>(impeller_scene_shaders_data,
+                                                 impeller_scene_shaders_length),
+          std::make_shared<fml::NonOwnedMapping>(
+              impeller_tessellation_shaders_data,
```
Comment (chinmaygarde) [inline impeller/playground/backend/metal/playground_impl_mtl.mm:55]: Nit: Consider naming these something other than tessellation shaders since these don't setup tessellation pipeline stages in the rasterizer.

File: impeller/renderer/BUILD.gn [hunk 6]
```diff
@@ -4,18 +4,84 @@
...
+  deps = [
+    "../base",
+    "../geometry",
+    "../image",
+    "//flutter/fml",
+  ]
+}
+
+if (impeller_enable_compute) {
+  impeller_shaders("tessellation_shaders") {
```
Comment (chinmaygarde) [inline impeller/renderer/BUILD.gn:44]: compute_shaders perhaps because of the overloading if the tessellation term?

File: impeller/renderer/BUILD.gn [hunk 7]
```diff
@@ -4,18 +4,84 @@
 
 import("//flutter/impeller/tools/impeller.gni")
 
-impeller_component("renderer") {
+impeller_component("allocation") {
```
Comment (chinmaygarde) [inline impeller/renderer/BUILD.gn:7]: This does more than just allocation and contains stuff like formats and descriptors. Perhaps `//impeller/core` or `//impeller/device`? (I like the former).

**Review by chinmaygarde — APPROVED** [review]

File: impeller/renderer/BUILD.gn (hunk 7 above)
Comment (dnfield) [reply impeller/renderer/BUILD.gn:7]: I was originally going to call this "buffers" but it also has textures and formats and an allocator. Right now it's a target in `//impeller/renderer` still. I'm fine naming it whatever, moving to core

File: impeller/renderer/BUILD.gn (hunk 6 above)
Comment (dnfield) [reply impeller/renderer/BUILD.gn:44]: done

File: impeller/playground/backend/metal/playground_impl_mtl.mm (hunk 5 above)
Comment (dnfield) [reply impeller/playground/backend/metal/playground_impl_mtl.mm:55]: done
